linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala

```diff
@@ -220,6 +220,38 @@ private[emitter] final class JSGen(val semantics: Semantics,
     })
   }
 
+  def genOriginalName(name: Name, originalName: OriginalName,
+      jsName: String): OriginalName = {
+    genOriginalName(name.encoded, originalName, jsName)
+  }
+
+  def genOriginalName(name: MethodName, originalName: OriginalName,
+      jsName: String): OriginalName = {
+    genOriginalName(name.simpleName, originalName, jsName)
+  }
+
+  private def genOriginalName(name: UTF8String, originalName: OriginalName,
+      jsName: String): OriginalName = {
+
+    def sameName: Boolean = {
+      // scalastyle:off return
+      if (name.length != jsName.length())
+        return false
+      var i = 0
+      while (i != name.length) {
+        if (name(i).toInt != jsName.charAt(i).toInt)
+          return false
+        i += 1
+      }
+      true
+      // scalastyle:on return
+    }
```

Comment:
Hum, this looks wrong. It compares an UTF16 string to an UTF8 string byte by byte...

IIUC, this is merely an optimization to *not* generate an `OriginalName` when it is definitely not necessary. If this is the case, please comment.